### PR TITLE
feat(node): Generate syncing related events

### PR DESCRIPTION
**Pull request by oblique**

### CODE REVIEW

File: node/src/events.rs [hunk 1]
```diff
@@ -272,6 +312,29 @@ impl fmt::Display for NodeEvent {
             NodeEvent::FatalDaserError { error } => {
                 write!(f, "Daser stopped because of a fatal error: {error}")
             }
+            NodeEvent::NewSubjectiveHead { height } => {
+                write!(f, "New subjective head: {height}")
+            }
+            NodeEvent::AddedHeaderFromHeaderSub { height } => {
+                write!(f, "Added header {height} from HeaderSub")
```
Comment (zvolin) [inline node/src/events.rs:319]:
```suggestion
                write!(f, "Added header {height} from header-sub")
```
I think protocol name uses this format, but up to u

File: node/src/events.rs [hunk 2]
```diff
@@ -219,6 +228,37 @@ pub enum NodeEvent {
...
+    AddedHeaderFromHeaderSub {
+        /// The height of the header.
+        height: u64,
+    },
+
+    /// Fetching header of network head in progress.
+    FetchingHeadHeader,
+
+    /// Fetching headers for a specific block range.
+    FetchingHeaders {
```
Comment (zvolin) [inline node/src/events.rs:248]: I think we should also have an event when we succeeded fetching and inserted them

File: cli/src/native.rs [hunk 3]
```diff
@@ -80,14 +80,15 @@ pub(crate) async fn run(args: Params) -> Result<()> {
...
-    let mut events = node.event_subscriber();
-
-    // We have nothing else to do, but we want to keep main alive
     while let Ok(ev) = events.recv().await {
         match ev.event {
             // Skip noisy events
             NodeEvent::ShareSamplingResult { .. } => continue,
+            event @ (NodeEvent::FatalDaserError { .. }
+            | NodeEvent::FetchingHeadersFailed { .. }
+            | NodeEvent::NetworkCompromised) => {
```
Comment (oblique) [inline cli/src/native.rs:89]: I was thinking if I should add a `is_error` method in `NodeEvent`. What do you think?

File: node/src/events.rs (hunk 2 above)
Comment (oblique) [reply node/src/events.rs]: Done

File: cli/src/native.rs (hunk 3 above)
Comment (zvolin) [reply cli/src/native.rs:89]: sure, why not

Comment (oblique) [reply cli/src/native.rs]: Done. Should we consider a rejected `ShareSamplingResult` an error and display it as a warning?

Comment (oblique) [reply cli/src/native.rs]: Also should I propagate the `is_error` to JS?

Comment (zvolin) [reply cli/src/native.rs]: I think yes to both

Comment (oblique) [reply cli/src/native.rs]: done

File: node/src/events.rs [hunk 4]
```diff
@@ -219,6 +228,82 @@ pub enum NodeEvent {
...
+    FetchingHeadersFinished {
+        /// Start of the range.
+        from_height: u64,
+        /// End of the range (included).
+        to_height: u64,
+        /// How much time fetching took.
+        took: Duration,
+    },
+
+    /// Fetching headers of a specific block range just finished.
```
Comment (fl0rek) [inline node/src/events.rs:267]:
```suggestion
    /// Fetching headers of a specific block range just failed.
```

File: node/src/syncer.rs [hunk 5]
```diff
@@ -419,6 +444,12 @@ where
             return;
         }
 
+        if self.p2p.peer_tracker_info().num_connected_peers == 0 {
+            // No connected peers. We can't do the request.
+            // This will be recovered by `run`.
```
Comment (fl0rek) [inline node/src/syncer.rs:449]:
```suggestion
            // We will recover from this in `run`.
```

**Review by fl0rek — COMMENTED** [review]
nice work 🙌

**Review by zvolin — APPROVED** [review]

**Review by fl0rek — APPROVED** [review]

**Pull request merged**